gadopt/level_set_tools.py

```diff
@@ -1,135 +1,247 @@
+      curve:
+        A string matching an implemented curve or a callable implementing a new curve
+      curve_args:
+        A tuple of arguments defining the curve
+
+    Returns:
+      A dictionary with the keyword arguments to initialise a signed-distance function
     """
 
-    density: Optional[Number] = None
-    B: Optional[Number] = None
-    RaB: Optional[Number] = None
+    def line(x, slope, intercept):
+        """Straight line equation"""
+        return slope * x + intercept
 
-    def __post_init__(self):
-        """Checks instance field values.
+    def cosine(x, amplitude, wavelength, vertical_shift, phase_shift=0):
+        """Cosine function with an amplitude and a vertical shift."""
+        cosine = np.cos(2 * np.pi / wavelength * x + phase_shift)
 
-        Raises:
-            ValueError:
-              Incorrect field types.
-        """
-        count_None = 0
-        for field_var in fields(self):
-            field_var_value = getattr(self, field_var.name)
-            if isinstance(field_var_value, Number):
-                self.density_B_RaB = field_var.name
-            elif field_var_value is None:
-                count_None += 1
+        return amplitude * cosine + vertical_shift
+
+    curves = {"line": line, "cosine": cosine}
+    if isinstance(curve, str):
+        curve = curves[curve]
+
+    interface_coords_y = curve(interface_coords_x, *curve_args)
+    interface_coords = np.column_stack((interface_coords_x, interface_coords_y))
+
+    signed_distance_kwargs = {
+        "interface_coordinates": interface_coords,
+        "interface_geometry": "curve",
+        "interface_callable": curve,
+        "interface_args": curve_args,
+    }
+
+    return signed_distance_kwargs
+
+
+def rectangle_interface(ref_vertex_coords: tuple[float], edge_sizes: tuple[float]):
+    """Material interface defined by a rectangle.
+
+    Edges are aligned with Cartesian directions and do not overlap domain boundaries.
+
+    Args:
+      ref_vertex_coords:
+        A tuple holding the coordinates of the lower-left vertex
+      edge_sizes:
+        A tuple holding the edge sizes
+
+    Returns:
+      A dictionary with the keyword arguments to initialise a signed-distance function
+    """
+    interface_coords = [
+        (ref_vertex_coords[0], ref_vertex_coords[1]),
+        (ref_vertex_coords[0] + edge_sizes[0], ref_vertex_coords[1]),
+        (ref_vertex_coords[0] + edge_sizes[0], ref_vertex_coords[1] + edge_sizes[1]),
+        (ref_vertex_coords[0], ref_vertex_coords[1] + edge_sizes[1]),
+        (ref_vertex_coords[0], ref_vertex_coords[1]),
+    ]
+
+    signed_distance_kwargs = {
+        "interface_coordinates": interface_coords,
+        "interface_geometry": "polygon",
+    }
+
+    return signed_distance_kwargs
+
+
+def signed_distance(
+    level_set: fd.Function,
+    /,
+    interface_coordinates: list[list[float]] | np.ndarray | list[tuple[float], float],
```

Comment:
As discussed in the demos, perhaps the first argument given could be *interface_geometry*. If it is *"curve"* then it looks at the *interface_args* which would include the *interface_x_coordinates*, callable/special curve like *cosine* etc and then uses *curve_interface* under the hood to get the *interface_coordinates*.

 Otherwise if polygon then you need to provide the *interface_coordinates* directly to *signed_distance*. Maybe this approach is simpler because it removes the extra *curve_interface* function in the demo which was only supposed to be a helper function to create the *interface_coordiantes*... What do you think?